jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/GracefulShutdownHandler.java

```diff
@@ -0,0 +1,161 @@
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.component.Graceful;
+import org.eclipse.jetty.util.statistic.CounterStatistic;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handler to track active requests and allow them to gracefully complete.
+ */
+public class GracefulShutdownHandler extends Handler.Wrapper implements Graceful
+{
+    private static final Logger LOG = LoggerFactory.getLogger(GracefulShutdownHandler.class);
+
+    private CounterStatistic dispatchedStats = new CounterStatistic();
+    private Shutdown shutdown;
+
+    class ShutdownTrackingCallback extends Callback.Nested
+    {
+        final Request request;
+        final Response response;
+
+        public ShutdownTrackingCallback(Request request, Response response, Callback callback)
+        {
+            super(callback);
+            this.request = request;
+            this.response = response;
+            dispatchedStats.increment();
+        }
+
+        public void decrement()
+        {
+            dispatchedStats.decrement();
+        }
+
+        @Override
+        public void succeeded()
+        {
+            decrement();
+            super.succeeded();
+            if (isShutdown())
+            {
+                shutdown.check();
+            }
+        }
+
+        @Override
+        public void failed(Throwable x)
+        {
+            decrement();
+
+            Callback callback = Callback.from(getCallback(), shutdown::check);
+
+            if (isShutdown())
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Service Unavailable: {}", request.getHttpURI(), x);
+                Response.writeError(request, response, callback, HttpStatus.SERVICE_UNAVAILABLE_503);
+            }
+            else
+            {
+                Response.writeError(request, response, callback, x);
+            }
+        }
+    }
+
+    private boolean serviceUnavailable(Request request, Response response, Callback callback)
+    {
+        if (!response.isCommitted())
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Service Unavailable (before wrapped process): {}", request.getHttpURI());
+            response.getHeaders().put(HttpHeader.CONNECTION, HttpHeaderValue.CLOSE);
+            Response.writeError(request, response, callback, HttpStatus.SERVICE_UNAVAILABLE_503);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean process(Request request, Response response, Callback callback) throws Exception
+    {
+        Handler handler = getHandler();
+        ShutdownTrackingCallback shutdownCallbackTracking = new ShutdownTrackingCallback(request, response, callback);
+        if (handler == null || !isStarted() || isShutdown())
+        {
+            // always return 503
+            if (serviceUnavailable(request, response, shutdownCallbackTracking));
+                return true;
+        }
```

Comment:
I don't  think we need to increment/decrement if we are already stopped or shutdown:

```suggestion
        if (handler == null || !isStarted() || isShutdown())
        {
            // always return 503
            if (serviceUnavailable(request, response, callback));
                return true;
        }
        ShutdownTrackingCallback shutdownCallbackTracking = new ShutdownTrackingCallback(request, response, callback);

```

Reply:
This change will prevent `shutdown.check()` behavior while in `isShutdown() == true`.
Is that acceptable?

Reply:
Actually ... shouldn't the error response be also part of the graceful shutdown?
In other words, shouldn't we let those finish before we allow the Connector or Endpoint to shutdown?

Reply:
@gregw your code has a race.

Thread1 enters, sees `isShutdown()==false`, but does not increment yet the counter.
Thread2 calls `shutdown()`, and the `CompletableFuture` is immediately completed because the counter is `0`, but that is wrong because Thread1 is now processing the request.

@joakime's version is first incrementing the counter, and then checking for `isShutdown()`, which avoids the race.
However, it deserves a comment about why the increment is always done as the first thing.

Reply:
The intention was that thread1 will never actually enter the processing of the request.   See better version in latest review.